napalm_srlinux/srlinux.py

```diff
@@ -0,0 +1,1036 @@
+    used to talk to the device.
+    """
+
+    class RPCMethod(str, enum.Enum):
+        """
+        Enum class used to represent RPC Methods
+        """
+
+        GET = "get"
+        SET = "set"
+        VALIDATE = "validate"
+        CLI = "cli"
+
+    class RPCAction(str, enum.Enum):
+        """
+        Enum class used to represent RPC Actions
+        """
+
+        REPLACE = "replace"
+        UPDATE = "update"
+        DELETE = "delete"
+
+    class Datastore(str, enum.Enum):
+        """
+        Enum class used to represent SR Linux Data stores for RPC calls.
+        """
+
+        CANDIDATE = "candidate"
+        RUNNING = "running"
+        STATE = "state"
+        TOOLS = "tools"
+
+    def __init__(
+        self,
+        hostname: str,
+        username: str,
+        password: str,
+        timeout: Optional[int] = 60,
+        optional_args: Optional[dict] = None,
+    ):
+        """Constructor."""
+        self.device = None
+        self.hostname = hostname
+        self.username = username
+        self.password = password
+        self.timeout = timeout
+
+        if optional_args is None:
+            optional_args = {}
+
+        # Optional Arguments
+        self.jsonrpc_port = self._determine_jsonrpc_port(optional_args)
+        self.skip_verify = optional_args.get("skip_verify", False)
+        self.insecure = optional_args.get("insecure", False)
+        self.tls_ca = optional_args.get("tls_ca", "")
+        self.tls_cert_path = optional_args.get("tls_cert_path", "")
+        self.tls_key_path = optional_args.get("tls_key_path", "")
+        self.tls_key_password = optional_args.get("tls_key_password", "")
+
+        self.jsonrpc_client = self._new_jsonrpc_client()
+        self.jsonrpc_url = self._compose_jsonrpc_url()
+
+        # Warn about incompatible/oddball settings
+        if self.jsonrpc_port == 80:
+            if not self.insecure:
+                logging.warning(
+                    "Secure JSON RPC uses port 443, not 80."
+                    + "Set 'insecure=True' flag to indicate this is ok"
+                )
+        elif self.jsonrpc_port != 443:
+            logging.warning(
+                f"Non-default JSON RPC port configured ({self.jsonrpc_port}), typically only 443(default) or 80 are used"
+            )
+
+        if not self.insecure:
+            if not self.tls_ca:
+                logging.warning(
+                    "Incompatible settings: insecure=False "
+                    + "requires certificate parameter 'tls_ca' to be set "
+                    + "when using self-signed certificates"
+                )
+
+    def open(self):
+        """Check the supplied init params actually work, throw an exception if not."""
+        try:
+            self.jsonrpc_client.head(self.jsonrpc_url)
+        except httpx.RequestError as exc:
+            raise ConnectionError(
+                "Error opening http(s) connection to JSON-RPC server"
+            ) from exc
```

Comment:
@tardoe I have rewritten this to use http.HEAD method instead of fetching the `version` field.
I think it is more clear, as it isn't subject to RBAC and purely tests that transport is working.

Reply:
Yep, great idea.